First simulation

KarolJastrzebski opened the pull request:
Contains setup and sample simulation for Gatling.

https://tools.hmcts.net/jira/browse/RPE-287

KarolJastrzebski committed:
Wireframe for Gatling tests

KarolJastrzebski committed:
Implement sample load simulation

Request token and check its validity multiple times.

Values for URL, service name and secret, and number of repetitions are
read from configuration file.

Parameters for simulation are subject to change.

lgonczar commented on src/gatling/scala/SingleLeaseMultipleChecks.scala:
```diff
@@ -0,0 +1,46 @@
+        exec(checkJwt)
+      }
+      .inject(
+        rampUsers(10) over (10 seconds),
```
What's the purpose of this config (ramp up to 10, hit with 100, do nothing, hit with 200?)? I guess we just need a ramp up to max users over N seconds and then keep running the simulation with max users for some time to see if it can handle the load. Probably with realistic pauses inside the scenario.

KarolJastrzebski replied:
Thanks Leszek, I am changing the plan.

lgonczar commented on src/gatling/scala/SingleLeaseMultipleChecks.scala:
```diff
@@ -0,0 +1,46 @@
+    http("Lease token")
+      .post("/lease")
+      .formParam("microservice", config.getString("service.name"))
+      .formParam("oneTimePassword", authenticator.getTotpPassword(config.getString("service.pass")))
```
I think this password will be used every time. We need to provide these passwords using a feeder.

KarolJastrzebski replied:
Some OTP are the same, however, they are generated with 30 seconds expiration, and tests can last longer than that.

lgonczar replied:
Yes, what I meant was that `authenticator.getTotpPassword(config.getString("service.pass"))` will be executed only once (before the simulations even start) and the password it returns will be used every time. At some point it may become invalid, as you said - that's why we need a feeder to generate a new password on the fly.

KarolJastrzebski committed:
Use feeder for generating TOTP

Allows running tests past 30-second limit.

KarolJastrzebski committed:
Simplify load scenario

lgonczar commented on src/gatling/scala/SingleLeaseMultipleChecks.scala:
```diff
@@ -0,0 +1,45 @@
+      .feed(otpFeeder)
+      .exec(requestJwt)
+      .repeat(config.getInt("repeats")) {
+        exec(checkJwt)
```
I'd add a short pause in this place to make it more similar to the real usage.

lgonczar commented on src/gatling/scala/SingleLeaseMultipleChecks.scala:
```diff
@@ -0,0 +1,45 @@
+        exec(checkJwt)
+      }
+      .inject(
+        rampUsers(10000) over (1 minute)
```
rampUsers is fine but the number is too great. We can be pretty sure s2s won't handle that many (probably Gatling won't be able to generate such load, either). Besides, we won't have 10000 instances of services hitting s2s. 50? 100?

lgonczar commented on src/gatling/scala/SingleLeaseMultipleChecks.scala:
```diff
@@ -0,0 +1,45 @@
+    scenario("Testing")
+      .feed(otpFeeder)
+      .exec(requestJwt)
+      .repeat(config.getInt("repeats")) {
```
It may be better to replace the repeats with a duration-based solution. We know that the token will become invalid within 30 seconds, but we don't know how many checks that will allow us to do.

Also, we're ramping users over 1 minute, but the scenarios can't take longer than 30 seconds (the expiry of the key), meaning we'll never see the maximum number of users hitting s2s simultaneously. In order to make the duration of a scenario less constrained, we need to request a new token every now and then (perhaps by introducing a top-level repeat?).

lgonczar reviewed (commented):
Have you checked it locally?

KarolJastrzebski committed:
Replace repeat with duration

Client checks continuously obtained JWT for 2 minutes with 200
millisecond pause.

Maximum number of users is generously set to 400.
